src/src/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -0,0 +1,359 @@
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.aad.adal;
+
+import java.io.IOException;
+
+import com.microsoft.aad.adal.AuthenticationRequest.UserIdentifierType;
+
+import android.content.Context;
+
+/**
+ * Internal class handling the detailed acquiretoken silent logic, including cache lookup and also
+ * interact with web request.
+ */
+class AcquireTokenSilentHandler {
+    private static final String TAG = AcquireTokenSilentHandler.class.getSimpleName();
+    
+    private final Context mContext;
+    private final TokenCacheAccessor mTokenCacheAccessor;
+    private final AuthenticationRequest mAuthRequest;
+    
+    private boolean mAttemptedWithFRT = false;
+    private TokenCacheItem mMrrtTokenCacheItem;
+
+    /**
+     * TODO: need to remove. {@link HttpUrlConnectionFactory} provides the possibility to 
+     * mock the real connection. Needs to update the class to make different response based on 
+     * post message. 
+     */
+    private IWebRequestHandler mWebRequestHandler = null;
+    
+    /**
+     * Constructor for {@link AcquireTokenSilentHandler}. 
+     * {@link TokenCacheAccessor} could be null. If null, won't handle with cache. 
+     */
+    AcquireTokenSilentHandler(final Context context, final AuthenticationRequest authRequest,
+            final TokenCacheAccessor tokenCacheAccessor) {
+        if (context == null) {
+            throw new IllegalArgumentException("context");
+        }
+        
+        if (authRequest == null) {
+            throw new IllegalArgumentException("authRequest");
+        }
+        
+        mContext = context;
+        mAuthRequest = authRequest;
+        mTokenCacheAccessor = tokenCacheAccessor;
+
+        mWebRequestHandler = new WebRequestHandler();
+    }
+    
+    /**
+     * Request for access token by looking up cache. . 
+     */
+    AuthenticationResult getAccessToken() throws AuthenticationException {
+        // If mTokenCacheAccessor is null, won't handle with token cache lookup. 
+        if (mTokenCacheAccessor == null) {
+            return null;
+        }
+        
+        TokenCacheItem tokenCacheItem = mTokenCacheAccessor.getRegularTokenCacheItemWithAT(mAuthRequest.getResource(), 
+                mAuthRequest.getClientId(), getUser());
+        
```

Comment:
If user is not passed , check for null user entry(return if there is only)-- note: for wildcard case, since RT represents a user, we should check if we have multiple RTs for same clientid, resource, authority
if user is passed, check for specific entries.